apollo-client/src/main/java/com/ctrip/framework/apollo/internals/K8sConfigMapConfigRepository.java

```diff
@@ -0,0 +1,301 @@
+        } catch (Throwable ex) {
+            Tracer.logEvent("ApolloConfigException", ExceptionUtil.getDetailMessage(ex));
+            transaction.setStatus(ex);
+            throw new ApolloConfigException("Create configmap failed!", ex);
+        } finally {
+            transaction.complete();
+        }
+    }
+
+    @Override
+    public Properties getConfig() {
+        if (configMapProperties == null) {
+            sync();
+        }
+        Properties result = propertiesFactory.getPropertiesInstance();
+        result.putAll(configMapProperties);
+        logger.info("configmap properties: {}", configMapProperties);
+        return result;
+    }
+
+    /**
+     * Update the memory when the configuration center changes
+     *
+     * @param upstreamConfigRepository the upstream repo
+     */
+    @Override
+    public void setUpstreamRepository(ConfigRepository upstreamConfigRepository) {
+        // 设置上游数据源
+        if (upstreamConfigRepository == null) {
+            return;
+        }
+        //clear previous listener
+        if (upstream != null) {
+            upstream.removeChangeListener(this);
+        }
+        upstream = upstreamConfigRepository;
+        upstreamConfigRepository.addChangeListener(this);
+    }
+
+    @Override
+    public ConfigSourceType getSourceType() {
+        return sourceType;
+    }
+
+    /**
+     * Sync the configmap
+     */
+    @Override
+    protected void sync() {
+        // Chain recovery, first read from upstream data source
+        boolean syncFromUpstreamResultSuccess = trySyncFromUpstream();
+
+        if (syncFromUpstreamResultSuccess) {
+            return;
+        }
+
+        Transaction transaction = Tracer.newTransaction("Apollo.ConfigService", "syncK8sConfigMap");
+        Throwable exception = null;
+        try {
+            configMapProperties = loadFromK8sConfigMap();
+            sourceType = ConfigSourceType.CONFIGMAP;
+            transaction.setStatus(Transaction.SUCCESS);
+        } catch (Throwable ex) {
+            Tracer.logEvent("ApolloConfigException", ExceptionUtil.getDetailMessage(ex));
+            transaction.setStatus(ex);
+            exception = ex;
+            throw new ApolloConfigException("Load config from Kubernetes ConfigMap failed!", ex);
+        } finally {
+            transaction.complete();
+        }
+
+        if (configMapProperties == null) {
+            sourceType = ConfigSourceType.NONE;
+            throw new ApolloConfigException(
+                    "Load config from Kubernetes ConfigMap failed!", exception);
+        }
+    }
+
+    public Properties loadFromK8sConfigMap() {
+        Preconditions.checkNotNull(configMapName, "ConfigMap name cannot be null");
+
+        try {
+            String jsonConfig = kubernetesManager.getValueFromConfigMap(k8sNamespace, configMapName, configMapKey);
+            if (jsonConfig == null) {
+                // TODO 这样重试访问idc，default是否正确
```

Comment:
You may refer the [documentation](https://www.apolloconfig.com/#/zh/design/apollo-introduction?id=_44-%e9%85%8d%e7%bd%ae%e8%8e%b7%e5%8f%96%e8%a7%84%e5%88%99) and the codes below.

https://github.com/apolloconfig/apollo/blob/ced6383de19b7cf4e37ab68ab8babcde7a9cbbba/apollo-configservice/src/main/java/com/ctrip/framework/apollo/configservice/service/config/AbstractConfigService.java#L40-L64